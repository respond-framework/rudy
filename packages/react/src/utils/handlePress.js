// @flow
import { redirect } from '@respond-framework/rudy'
import type { Routes, ReceivedAction } from '@respond-framework/rudy'

export type OnClick = false | ((SyntheticEvent) => ?boolean)
export default (
  action: ?ReceivedAction,
  routes: Routes,
  shouldDispatch: boolean,
  dispatch: Function,
  onClick?: ?OnClick,
  target: ?string,
  isRedirect?: boolean,
  fullUrl: string,
  history: Object,
  e: SyntheticEvent,
) => {
  const prevented = e.defaultPrevented
  const notModified = !isModified(e)
  let shouldGo = true

  if (onClick) {
    shouldGo = onClick(e) !== false // onClick can return false to prevent dispatch
  }

  if (!target && e && e.preventDefault && notModified) {
    e.preventDefault()
  }

  if (
    action &&
    shouldGo &&
    shouldDispatch &&
    !target &&
    !prevented &&
    notModified &&
    e.button === 0
  ) {
    action = isRedirect ? redirect(action) : action
    return dispatch(action)
  }

  if (!action && !target && fullUrl.indexOf('http') === 0) {
    window.location.href = fullUrl
  }
  return undefined
}

const isModified = (e: Object) =>
  !!(e.metaKey || e.altKey || e.ctrlKey || e.shiftKey)
