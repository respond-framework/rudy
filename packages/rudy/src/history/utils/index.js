export { supportsDom } from './supports'

export {
  addPopListener,
  removePopListener,
  isExtraneousPopEvent,
} from './popListener'

export {
  saveHistory,
  restoreHistory,
  pushState,
  replaceState,
} from './sessionStorage'
