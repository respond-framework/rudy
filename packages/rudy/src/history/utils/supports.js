// @flow
/* global window */

// eslint-disable-next-line import/prefer-default-export
export const supportsDom = () =>
  !!(
    typeof window !== 'undefined' &&
    window.document &&
    window.document.createElement
  )
