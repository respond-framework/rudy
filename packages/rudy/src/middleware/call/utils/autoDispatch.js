export default (req, callback, route, name, isOptCb) =>
  Promise.resolve(callback(req)).then((res) =>
    tryDispatch(req, res, route, name, isOptCb),
  )

const tryDispatch = (req, res, route, name, isOptCb) => {
  if (res === false) return false
  const hasReturn = res === null || (res && !res._dispatched) // `res._dispatched` indicates it was manually dispatched

  if (hasReturn && isAutoDispatch(route, req.options, isOptCb)) {
    // if no dispatch was detected, and a result was returned, dispatch it automatically
    return Promise.resolve(req.dispatch(res))
  }

  return res
}

const isAutoDispatch = (route, options, isOptCb) =>
  isOptCb
    ? options.autoDispatch === undefined
      ? true
      : options.autoDispatch
    : route.autoDispatch !== undefined
    ? route.autoDispatch
    : options.autoDispatch === undefined
    ? true
    : options.autoDispatch
