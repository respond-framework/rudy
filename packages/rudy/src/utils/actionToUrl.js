// @flow
import pathToRegexp from 'path-to-regexp'
import { applyStringDefault, applyObjectDefault } from './applyDefaults'
import { compileUrl, cleanBasename } from './index'

import type {
  Route,
  Routes,
  ReceivedAction,
  Options,
  ToPath,
} from '../flow-types'

export default (
  action: ReceivedAction,
  api: Object,
  prevRoute?: string,
): Object => {
  const { routes, options: opts }: Object = api
  const { type, params, query, state, hash, basename }: ReceivedAction = action

  const route = routes[type] || {}
  const path: string | void = typeof route === 'object' ? route.path : route

  const p: Object = formatParams(params, route, opts) || {}
  const q: ?Object = formatQuery(query, route, opts)
  const s: Object = formatState(state, route, opts) || {}
  const h: string = formatHash(hash || '', route, opts)

  const bn = cleanBasename(basename)
  const isWrongBasename = bn && !opts.basenames.includes(bn)
  if (basename === '') s._emptyBn = true // not cool kyle

  try {
    if (isWrongBasename) {
      throw new Error(`[rudy] basename "${bn}" not in options.basenames`)
    }

    // path-to-regexp throws for failed compilations
    const pathname: string = compileUrl(path, p, q, h, route, opts) || '/'
    const url: string = bn + pathname

    return { url, state: s }
  } catch (e) {
    if (process.env.NODE_ENV === 'development') {
      // eslint-disable-next-line no-console
      console.error(
        `[rudy] unable to compile action "${type}" to URL`,
        action,
        e,
      )
    } else if (process.env.NODE_ENV === 'test') {
      // eslint-disable-next-line no-console
      console.log(`[rudy] unable to compile action "${type}" to URL`, action, e)
    }

    const base: string = isWrongBasename ? '' : bn
    const url: string =
      base + notFoundUrl(action, routes, opts, q, h, prevRoute)
    return { url, state: s }
  }
}

const formatParams = (
  params: ?Object,
  route: Route,
  opts: Options,
): void | {} => {
  params = applyObjectDefault(route.defaultParams || opts.defaultParams)(
    params,
    route,
    opts,
  )

  if (params) {
    const keys = []
    pathToRegexp(typeof route === 'string' ? route : route.path, keys)
    const newParams: {} = {}
    const toPath: ToPath = route.toPath || opts.toPath || defaultToPath
    keys.forEach(
      ({
        name,
        repeat,
        optional,
      }: {
        name: string | Number,
        repeat: boolean,
        optional: boolean,
      }) => {
        if (!Object.prototype.hasOwnProperty.call(params, name)) {
          return
        }
        // $FlowFixMe
        const val = params[name]
        const urlVal = toPath(
          val,
          { name: name.toString(), repeat, optional },
          route,
          opts,
        )
        if (repeat) {
          if (!Array.isArray(urlVal)) {
            throw Error('toPath failed')
          }
          if (!optional && !urlVal.length) {
            throw Error('toPath failed')
          }
          urlVal.forEach((segment) => {
            if (typeof segment !== 'string' || !segment) {
              throw Error('toPath failed')
            }
          })
        } else if (
          typeof urlVal !== 'string' &&
          (!optional || urlVal !== undefined)
        ) {
          throw Error('toPath failed')
        }
        newParams[name.toString()] = urlVal
      },
    )
    return newParams
  }
  return undefined
}

const toSegment = (val, convertNum, capitalize, optional) => {
  if (typeof val === 'number' && convertNum) {
    return val.toString()
  }
  if (typeof val !== 'string' || (optional && !val)) {
    throw TypeError('[rudy]: defaultToPath::toSegment received unknown type')
  }
  if (capitalize) {
    return val.replace(/ /g, '-').toLowerCase()
  }
  return val
}

export const defaultToPath: ToPath = (
  val,
  { repeat, optional },
  route,
  opts,
) => {
  const convertNum =
    route.convertNumbers ||
    (opts.convertNumbers && route.convertNumbers !== false)

  const capitalize =
    route.capitalizedWords ||
    (opts.capitalizedWords && route.capitalizedWords !== false)

  if (repeat) {
    if (optional && val === undefined) {
      return []
    }
    if (!val) {
      throw Error('defaultToPath received incorrect value')
    }
    return val.split('/')
  }
  if (optional && val === undefined) {
    return undefined
  }
  return toSegment(val, convertNum, capitalize, optional)
}

export const formatQuery = (
  query: ?Object,
  route: Route,
  opts: Options,
): ?Object => {
  query = applyObjectDefault(route.defaultQuery || opts.defaultQuery)(
    query,
    route,
    opts,
  )
  const to = route.toSearch || opts.toSearch

  if (to && query) {
    const newQuery = {}

    Object.keys(query).forEach((key) => {
      // $FlowFixMe
      newQuery[key] = to(query[key], key, route, opts)
    })

    return newQuery
  }

  return query
}

export const formatHash = (
  hash: string,
  route: Route,
  opts: Options,
): string => {
  hash = applyStringDefault(route.defaultHash || opts.defaultHash)(
    hash,
    route,
    opts,
  )
  const to = route.toHash || opts.toHash
  return to ? to(hash, route, opts) : hash
}

const formatState = (state: ?Object, route: Route, opts: Options): ?Object =>
  applyObjectDefault(route.defaultState || opts.defaultState)(
    state,
    route,
    opts,
  )
// state has no string counter part in the address bar, so there is no `toState`

const notFoundUrl = (
  action: ReceivedAction,
  routes: Routes,
  opts: Options,
  query: ?Object,
  hash: string,
  prevRoute: string = '',
): string => {
  const type: string = action.type || ''
  const route: {} = routes[type] || {}
  const hasScene: boolean = type.indexOf('/NOT_FOUND') > -1
  // TODO: Look into scene stuff
  // $FlowFixMe
  const scene: string = route.scene || prevRoute.scene || ''
  const t: string = hasScene
    ? type
    : routes[`${scene}/NOT_FOUND`] // try to interpret scene-level NOT_FOUND if available (note: links create plain NOT_FOUND actions)
    ? `${scene}/NOT_FOUND`
    : 'NOT_FOUND'

  const p: string = routes[t].path || routes.NOT_FOUND.path || ''

  // preserve these (why? because we can)
  const s: string =
    query && opts.stringifyQuery ? `?${opts.stringifyQuery(query)}` : ''
  const h: string = hash ? `#${hash}` : ''

  return p + s + h
}
