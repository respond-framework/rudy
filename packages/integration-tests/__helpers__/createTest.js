import { applyMiddleware, createStore, combineReducers } from 'redux'
import {
  get,
  clear,
} from '@respond-framework/rudy/history/utils/sessionStorage'
import { locationToUrl } from '@respond-framework/rudy/utils'
import { createRouter } from '@respond-framework/rudy'
import awaitUrlChange from './awaitUrlChange'

export default async (...allArgs) => {
  const args = allArgs.filter((arg) => typeof arg !== 'function')
  const callbacks = allArgs.filter((arg) => typeof arg === 'function')
  const hasCallbacks = callbacks.length > 0

  const [testName, routesMap] = args
  let [, , options = {}, actions] = args

  if (Array.isArray(options)) {
    actions = options
    options = {}
  }

  options = Object.assign({}, JSON.parse(process.env.RUDY_OPTIONS), options) // do things like force all tests to log -- see ../.testsConfig.json

  options.wallabyErrors =
    options.wallabyErrors !== undefined
      ? options.wallabyErrors
      : process.env.WALLABY === 'true'

  if (options.skip) return

  options.testBrowser = options.testBrowser || false

  if (!actions) {
    actions = Object.keys(routesMap).filter(
      (type) => !/FIRST|REDIRECTED/.test(type),
    )
  }

  const initialPath =
    typeof actions[0] === 'string' && actions[0].charAt(0) === '/'
      ? actions.shift()
      : '/first'

  const hasMultipleTests = actions.length > 1
  let num = 1

  if (hasMultipleTests) {
    describe(testName, () => {
      for (const action of actions) {
        if (Array.isArray(action)) {
          const name = action[0]
          const act = action[1]
          createTest(name, routesMap, initialPath, act, options, num++)
        } else {
          const name = JSON.stringify(action)
          createTest(name, routesMap, initialPath, action, options, num++)
        }
      }

      if (hasCallbacks) {
        callbacks.forEach((cb, index) => {
          const name = callbacks.length === 1 ? 'snipes' : `snipes ${++index}`
          createSnipes(name, routesMap, initialPath, options, cb)
        })
      }
    })
  } else {
    const isSnipesOnly = initialPath === '/first' && actions.length === 0

    if (hasCallbacks && isSnipesOnly) {
      callbacks.forEach((cb, index) => {
        const name = callbacks.length === 1 ? testName : `snipes ${++index}`
        createSnipes(name, routesMap, initialPath, options, cb)
      })
    } else if (hasCallbacks) {
      describe(testName, () => {
        const name = actions[0]
          ? JSON.stringify(actions[0])
          : `firstRoute - ${initialPath}`

        createTest(name, routesMap, initialPath, actions[0], options, num)

        callbacks.forEach((cb, index) => {
          const name = callbacks.length === 1 ? 'snipes' : `snipes ${++index}`
          createSnipes(name, routesMap, initialPath, options, cb)
        })
      })
    } else if (Array.isArray(actions[0])) {
      const name = actions[0][0]
      const act = actions[0][1]

      createTest(name, routesMap, initialPath, act, options, num)
    } else {
      createTest(testName, routesMap, initialPath, actions[0], options, num)
    }
  }
}

const createTest = (testName, routesMap, initialPath, item, opts, num) => {
  test(testName, async () => {
    const {
      store,
      history,
      routes,
      options,
      firstRoute,
      initialState,
    } = setupStore(routesMap, initialPath, opts)

    const firstAction = firstRoute(false)
    const res = await store.dispatch(firstAction)

    if (routesMap.FIRST || initialPath !== '/first') {
      const prefix = `firstRoute - ${initialPath} - ${num}`
      snapChange(prefix, res, store, history, opts, initialState)
    }

    if (typeof item === 'string' && item.charAt(0) === '/') {
      const url = item
      const res = await history.push(url)

      snapChange(num, res, store, history, opts)
    } else if (item) {
      const action = typeof item === 'string' ? { type: item } : item
      const res = await store.dispatch(action)

      snapChange(num, res, store, history, opts)
    }

    snapRoutes(num, routes)
    snapOptions(num, options)

    if (opts.log) {
      if (typeof opts.log !== 'function') {
        console.log(store.getState().location)
      } else {
        opts.log(store.getState().location)
      }
    }

    history.unlisten()
  })
}

const createSnipes = (testName, routesMap, initialPath, opts, callback) => {
  test(testName, async () => {
    const { store, history, routes, options, firstRoute } = setupStore(
      routesMap,
      initialPath,
      opts,
    )

    const firstResponse =
      opts.dispatchFirstRoute === false
        ? null
        : await store.dispatch(firstRoute(false))

    const pop = opts.testBrowser && createPop(history)
    let defaultPrefix = 0

    await callback({
      firstRoute,
      firstResponse,
      history,
      routes,
      options,
      store,
      pop,
      dispatch: store.dispatch,
      getState: store.getState,
      getLocation: () => store.getState().location,
      snapChange: (prefix, res) => {
        if (opts.snipesOnly) return res

        if (typeof prefix === 'string') {
          return snapChange(prefix, res, store, history, opts)
        }

        res = prefix
        return snapChange(++defaultPrefix, res, store, history, opts)
      },
      snap: async (action, prefix = '') => {
        prefix = prefix || JSON.stringify(action) || 'function'

        const res = await store.dispatch(action)

        if (opts.snipesOnly) return res

        snapChange(prefix, res, store, history, opts)
        snapRoutes(prefix, routes)
        snapOptions(prefix, options)

        return res
      },
      snapPop: async (direction, prefix = '') => {
        const res = await pop(direction)

        if (opts.snipesOnly) return res

        snapChange(prefix, res, store, history, opts)
        snapRoutes(prefix, routes)
        snapOptions(prefix, options)

        return res
      },
    })

    if (opts.log) {
      if (typeof opts.log !== 'function') {
        console.log(store.getState().location)
      } else {
        opts.log(store.getState().location)
      }
    }

    history.unlisten()
  })
}

export const setupStore = (routesMap, initialPath, opts) => {
  const routes = createRoutes(routesMap)
  const options = createOptions(opts)

  options.initialEntries = [initialPath]
  options.extra = options.extra || { arg: 'extra-arg' }

  const middlewareFunc = options.middlewareFunc
  delete options.middlewareFunc

  const title = (state, action = {}) => {
    if (action.type.indexOf('@@redux') === 0) {
      return state || '<initial_title>'
    }
    return action.payload !== undefined
      ? `${action.type} - ${JSON.stringify(action.payload)}`
      : action.type
  }

  const {
    middleware,
    reducer,
    firstRoute,
    api: { history },
  } = createRouter(routes, options, middlewareFunc)

  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware)
  const store = createStore(rootReducer, enhancer)

  const initialState = store.getState()

  return {
    store,
    firstRoute,
    initialState,
    routes,
    options,
    history,
  }
}

const callbacks = [
  'beforeLeave',
  'beforeEnter',
  'onLeave',
  'onLeave',
  'onEnter',
  'thunk',
  'onComplete',
  'onError',
]

const createRoutes = (routesMap) => {
  const routes = {}

  for (const type in routesMap) {
    routes[type] =
      typeof routesMap[type] === 'object'
        ? { ...routesMap[type] }
        : routesMap[type]

    const route = routes[type]

    for (const cb in route) {
      if (typeof route[cb] === 'function') {
        route[cb] = jest.fn(route[cb])
      }
    }
  }

  // we add them like this (rather than merging over them) to preserve natural order
  // the tests would like the routes defined in. Order matters for matching types to URLs.
  // This essentially replicates what would typically happen in userland where they aren't
  // override routes for the convenience of tests.
  if (!routes.FIRST) {
    routes.FIRST = {
      path: '/first',
    }
  }

  if (!routes.NEVER_USED_PATHLESS) {
    routes.NEVER_USED_PATHLESS = {
      thunk: jest.fn(),
    }
  }

  if (!routes.REDIRECTED) {
    routes.REDIRECTED = {
      path: '/redirected',
      beforeEnter: () => new Promise((res) => setTimeout(res, 1)),
      onComplete: jest.fn(() => 'redirect_complete'),
    }
  }

  return routes
}

const createOptions = (opts = {}) => {
  const options = { ...opts }

  for (const cb in options) {
    if (typeof opts[cb] === 'function') {
      options[cb] = jest.fn(opts[cb])
    }
  }

  return options
}

const snapRoutes = (prefix, routes) => {
  for (const type in routes) {
    const route = routes[type]
    snapCallbacks(`${prefix} - routes - ${type}`, route)
  }
}

const snapOptions = (prefix, options) => {
  snapCallbacks(`${prefix} - options`, options)
}

const snapCallbacks = (prefix, obj) => {
  expectBeforeLeave(prefix, obj)
  expectBeforeEnter(prefix, obj)
  expectOnLeave(prefix, obj)
  expectOnEnter(prefix, obj)
  expectThunk(prefix, obj)
  expectOnComplete(prefix, obj)
  expectOnError(prefix, obj)

  snapOtherFunctions(prefix, obj)
}

// snapshot the number of calls of all additional functions that aren't middleware callbacks
const snapOtherFunctions = (prefix, obj) => {
  for (const k in obj) {
    if (!callbacks.includes(k) && typeof obj[k] === 'function' && obj[k].mock) {
      expect(obj[k].mock.calls.length).toMatchSnapshot(`${prefix} - ${k}`)
    }
  }
}

const snapChange = (prefix, res, store, history, opts = {}, initialState) => {
  if (initialState) expectInitialState(prefix, initialState)

  expectState(prefix, store)
  expectResponse(prefix, res)
  expectTitle(prefix)

  if (opts.testBrowser) {
    expectSessionStorage(prefix)
    expectWindowLocation(prefix)
  }
}

// all these expect functions are broken out separately so we can easily see the
// name of the expectation that failed in the trace displayed in the Wallaby web panel

const expectInitialState = (prefix, initialState) => {
  expect(initialState).toMatchSnapshot(`${prefix} - initial_state`)
}

const expectState = (prefix, store) => {
  expect(store.getState()).toMatchSnapshot(`${prefix} - state`)
}

const expectResponse = (prefix, res) => {
  expect(res).toMatchSnapshot(`${prefix} - response`)
}

const expectTitle = (prefix) => {
  expect(document.title).toMatchSnapshot(`${prefix} - title`)
}

const expectSessionStorage = (prefix) => {
  expect(get()).toMatchSnapshot(`${prefix} - sessionStorage`)
}

const expectWindowLocation = (prefix) => {
  expect(locationToUrl(window.location)).toMatchSnapshot(
    `${prefix} - windowLocation`,
  )
}

const expectBeforeLeave = (prefix, obj) => {
  if (typeof obj.beforeLeave === 'function' && obj.beforeLeave.mock) {
    expect(obj.beforeLeave.mock.calls.length).toMatchSnapshot(
      `${prefix} - beforeLeave`,
    )
  }
}

const expectBeforeEnter = (prefix, obj) => {
  if (typeof obj.beforeEnter === 'function' && obj.beforeEnter.mock) {
    expect(obj.beforeEnter.mock.calls.length).toMatchSnapshot(
      `${prefix} - beforeEnter`,
    )
  }
}

const expectOnLeave = (prefix, obj) => {
  if (typeof obj.onLeave === 'function' && obj.onLeave.mock) {
    expect(obj.onLeave.mock.calls.length).toMatchSnapshot(`${prefix} - onLeave`)
  }
}

const expectOnEnter = (prefix, obj) => {
  if (typeof obj.onEnter === 'function' && obj.onEnter.mock) {
    expect(obj.onEnter.mock.calls.length).toMatchSnapshot(`${prefix} - onEnter`)
  }
}

const expectThunk = (prefix, obj) => {
  if (typeof obj.thunk === 'function' && obj.thunk.mock) {
    expect(obj.thunk.mock.calls.length).toMatchSnapshot(`${prefix} - thunk`)
  }
}

const expectOnComplete = (prefix, obj) => {
  if (typeof obj.onComplete === 'function' && obj.onComplete.mock) {
    expect(obj.onComplete.mock.calls.length).toMatchSnapshot(
      `${prefix} - onComplete`,
    )
  }
}

const expectOnError = (prefix, obj) => {
  if (typeof obj.onError === 'function' && obj.onError.mock) {
    expect(obj.onError.mock.calls.length).toMatchSnapshot(`${prefix} - onError`)
  }
}

// for testing pops in a "real" browser (as simulated by Jest)

const createPop = (history) => async (direction = 'back') => {
  history.currentPop = null
  window.history[direction]()
  return awaitPop(history)
}

const awaitPop = async (history, tries = 1) => {
  if (tries >= 10) {
    throw new Error(`awaitPop reached the maximum amount of tries (${tries})`)
  }

  await new Promise((res) => setTimeout(res, 3))
  return history.currentPop || awaitPop(history, ++tries)
}

// needed to put multiple browser tests in the same file

export const resetBrowser = async () => {
  /* eslint-env browser */
  const storage = get()
  const index = storage && storage.index

  if (index) {
    const delta = index * -1
    const oldUrl = locationToUrl(window.location)

    window.history.go(delta)

    if (oldUrl !== '/') {
      // will otherwise fail because the last route was also `/` (all browser tests start on `/`)
      await awaitUrlChange() // we need for `go` to complete
    }
  }

  window.history.replaceState({}, null, '/') // insure if the index was 0 but there was a replace on it, that we are back at '/'
  clear()
  /* eslint-env */
}
