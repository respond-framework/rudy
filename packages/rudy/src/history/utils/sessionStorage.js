// @flow
/* eslint-env browser */
import { toEntries } from '../../utils'

// API:

// Below is the facade around both `sessionStorage` and our "history as storage" fallback.
//
// - `saveHistory` is  called every time the history entries or index changes
// - `restoreHistory` is called on startup obviously

// Essentially the idea is that if there is no `sessionStorage`, we maintain the entire
// storage object on EACH AND EVERY history entry's `state`. I.e. `history.state` on
// every page will have the `index` and `entries` array. That way when browsers disable
// cookies/sessionStorage, we can still grab the data we need off off of history state :)
//
// It's a bit crazy, but it works very well, and there's plenty of space allowed for storing
// things there to get a lot of mileage out of it. We store the minimum amount of data necessary.
//
// Firefox has the lowest limit of 640kb PER ENTRY. IE has 1mb and chrome has at least 10mb:
// https://stackoverflow.com/questions/6460377/html5-history-api-what-is-the-max-size-the-state-object-can-be

export const saveHistory = ({ entries }) => {
  entries = entries.map((e) => [e.location.url, e.state, e.location.key]) // one entry has the url, a state object, and a 6 digit key
  set({ entries })
}

export const restoreHistory = (api) => {
  let history = get()
  if (!history) {
    history = initializeHistory()
  } else {
    history.index = getHistoryState().index
  }
  return format(history, api)
}

export const clear = () => {
  window.sessionStorage.setItem(key(), '')
  historySet({ index: 0, id: key() })
}

const set = (val) => window.sessionStorage.setItem(key(), JSON.stringify(val))

export const get = () => {
  try {
    const json = window.sessionStorage.getItem(key())
    return JSON.parse(json)
  } catch (error) {
    return null
  }
}

// HISTORY FACADE:

export const pushState = (url: string) =>
  window.history.pushState(
    { id: sessionId(), index: getHistoryState().index + 1 },
    null,
    url,
  ) // insure every entry has the sessionId (called by `BrowserHistory`)

export const replaceState = (url: string) =>
  window.history.replaceState(
    { id: sessionId(), index: getHistoryState().index },
    null,
    url,
  ) // QA: won't the fallback overwrite the `id`? Yes, but the fallback doesn't use the `id` :)

export const go = (n) => window.history.go(n)

export const getCurrentIndex = () => getHistoryState().index

const historySet = (history) => window.history.replaceState(history, null) // set on current entry

// SESSION STORAGE FACADE:

// We use `history.state.id` to pick which "session" from `sessionStorage` to use in
// the case that multiple instances of the app exist in the browser history stack of
// the same tab (e.g. if you navigate away from the app and then back again)
let _id

const PREFIX = '@@rudy/'

const sessionId = () => (_id = _id || createSessionId())

const key = () => PREFIX + sessionId()

const createSessionId = () => {
  const state = getHistoryState()

  if (!state.id) {
    if (process.env.NODE_ENV === 'test') {
      state.id = '123456789'.toString(36).substr(2, 6)
    } else {
      state.id = Math.random()
        .toString(36)
        .substr(2, 6)
    }
    state.index = 0
    historySet(state)
  }

  return state.id
}

// HELPERS:

const initializeHistory = () => {
  const { pathname, search, hash } = window.location
  const url = pathname + search + hash
  return { n: 1, index: 0, entries: [url] } // default history on first load
}

const format = (history, api) => {
  const { entries, index } = history
  return toEntries(api, entries, index)
}

// IE11 sometimes throws when accessing `history.state`:
//
// - https://github.com/ReactTraining/history/pull/289
// - https://github.com/ReactTraining/history/pull/230#issuecomment-193555362
//
// The issue occurs:
// A) when you refresh a page that is the only entry and never had state set on it,
// which means it wouldn't have any state to remember in the first place
//
// B) in IE11 on load in iframes, which also won't need to remember state, as iframes
// usually aren't for navigating to other sites (and back). This may just be issue A)
const getHistoryState = () => {
  try {
    return window.history.state || {}
  } catch (e) {
    return {}
  }
}
