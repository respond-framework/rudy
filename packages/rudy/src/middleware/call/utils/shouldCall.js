// @flow
import { isServer } from '@respond-framework/utils'
import type { LocationState } from '../../../flow-types'
import { isHydrate } from '../../../utils'

export default (name, route, req, { runOnServer, runOnHydrate }) => {
  if (!route[name] && !req.options[name]) return false

  if (isHydrate(req) && !runOnHydrate) return false

  if (isServer() && !runOnServer) return false

  return allowBoth
}

const allowBoth = { route: true, options: true }

// If for instance, you wanted to allow each route to decide
// whether to skip options callbacks, here's a simple way to do it:
//
// return {
//   options: !route.skipOpts, // if true, don't make those calls
//   route: true
// }
//
// You also could choose to automatically trigger option callbacks only as a fallback:
//
// return {
//   options: !route[name],
//   route: !!route[name]
// }
