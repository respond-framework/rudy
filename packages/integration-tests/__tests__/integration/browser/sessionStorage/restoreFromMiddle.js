import { get } from '@respond-framework/rudy/history/utils/sessionStorage'
import createTest, { setupStore } from '../../../../__helpers__/createTest'

beforeAll(async () => {
  const routesMap = {
    FIRST: '/',
    SECOND: '/second',
    THIRD: '/third',
  }

  const { store, firstRoute, history } = setupStore(routesMap)

  const firstAction = firstRoute(false)
  await store.dispatch(firstAction)

  await store.dispatch({ type: 'SECOND' })
  await store.dispatch({ type: 'THIRD' })
  await history.back() // go back, so we can simulate leaving in the middle of the stack

  history.unlisten()
})

createTest(
  'restore history when index > 0',
  {
    FIRST: '/',
    SECOND: '/second',
    THIRD: '/third',
  },
  { testBrowser: true },
  [],
  async ({ snapPop, getLocation }) => {
    expect(getLocation()).toMatchSnapshot()
    expect(get()).toMatchSnapshot()

    await snapPop('back')

    expect(getLocation().type).toEqual('FIRST')
    expect(window.location.pathname).toEqual('/')

    expect(getLocation().index).toEqual(0)

    // although we left in the middle of the entries (at SECOND),
    // THIRD will still be present on return, since in a real browser
    // there is no reliable way to distinguish between these histories
    //  1. FIRST -> SECOND -> THIRD -> SECOND(back button) -> google.com -> SECOND(back button)
    //  2. FIRST -> SECOND -> THIRD -> google.com -> SECOND(right click back button, back 2 steps)
    expect(getLocation().length).toEqual(3)

    expect(get()).toMatchSnapshot()
  },
)
