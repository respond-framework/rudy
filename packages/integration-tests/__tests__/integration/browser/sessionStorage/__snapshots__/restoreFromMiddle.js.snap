// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`restore history when index > 0 1`] = `
Object {
  "basename": "",
  "blocked": null,
  "direction": "forward",
  "entries": Array [
    Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
  ],
  "from": null,
  "hash": "",
  "index": 1,
  "key": "345678",
  "kind": "load",
  "length": 2,
  "n": 1,
  "params": Object {},
  "pathname": "/second",
  "pop": false,
  "prev": Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "index": 0,
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
  "query": Object {},
  "scene": "",
  "search": "",
  "state": Object {},
  "status": 200,
  "type": "SECOND",
  "universal": false,
  "url": "/second",
}
`;

exports[`restore history when index > 0 2`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`restore history when index > 0 3`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`restore history when index > 0:  - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "blocked": null,
    "direction": "backward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
    ],
    "from": null,
    "index": 0,
    "key": "345678",
    "kind": "back",
    "length": 2,
    "n": -1,
    "pathname": "/",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "FIRST",
}
`;

exports[`restore history when index > 0:  - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`restore history when index > 0:  - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`restore history when index > 0:  - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/second",
      Object {},
      "345678",
    ],
  ],
}
`;

exports[`restore history when index > 0:  - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "backward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/second",
          "scene": "",
          "search": "",
          "url": "/second",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "SECOND",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "back",
    "length": 2,
    "n": -1,
    "params": Object {},
    "pathname": "/",
    "pop": true,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "FIRST",
}
`;

exports[`restore history when index > 0:  - title 1`] = `"FIRST"`;

exports[`restore history when index > 0:  - windowLocation 1`] = `"/"`;
